stl/inc/shared_mutex

```diff
@@ -227,46 +279,85 @@ public:
 
     shared_lock() noexcept = default;
 
-    _NODISCARD_CTOR_LOCK explicit shared_lock(mutex_type& _Mtx)
+    _NODISCARD_CTOR_LOCK
+    _Acquires_shared_lock_(_Mtx)
+    _Acquires_nonreentrant_lock_(_Mtx)
+    _Post_same_lock_(_Mtx, *this->_Pmtx)
+    _Post_satisfies_(_Owns == true)
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    explicit shared_lock(mutex_type& _Mtx)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(true) { // construct with mutex and lock shared
         _Mtx.lock_shared();
     }
 
+    _Post_satisfies_(_Owns == false)
+    _Post_satisfies_(_Pmtx == &_Mtx)
     shared_lock(mutex_type& _Mtx, defer_lock_t) noexcept
         : _Pmtx(_STD addressof(_Mtx)), _Owns(false) {} // construct with unlocked mutex
 
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, try_to_lock_t)
+    _NODISCARD_CTOR_LOCK
+    _Requires_lock_not_held_(_Mtx)
+    _When_(this->_Owns == true, _Acquires_shared_lock_(_Mtx))
+    _When_(this->_Owns == true, _Acquires_nonreentrant_lock_(_Mtx))
+    _When_(this->_Owns == true, _Post_same_lock_(_Mtx, *this->_Pmtx))
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, try_to_lock_t)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(_Mtx.try_lock_shared()) {} // construct with mutex and try to lock shared
 
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, adopt_lock_t) noexcept // strengthened
+    _NODISCARD_CTOR_LOCK
+    _Requires_lock_held_(_Mtx)

```

Comment:
For `shared_lock(mutex_type& _Mtx, adopt_lock_t)`, this says `_Requires_lock_held_`, but I think it should specifically be `_Requires_shared_lock_held_`. We're assuming that `_Mtx.lock_shared();` has already been called, when the plain ctor does so it `_Acquires_shared_lock_`, and the dtor is calling `unlock_shared()` and is annotated with `_Releases_shared_lock_`.

Reply:
Good point. Thanks.